test/Test/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -178,24 +188,67 @@ modelFoldMergingTree = goMergingTree
         OngoingTreeMergeData mr ->
           goMergingRun mr
         PendingLevelMergeData prs t ->
-          modelMerge MergeLevel (map goPreExistingRun prs <> map goMergingTree (toList t))
+          modelMerge MR.MergeLevel (map goPreExistingRun prs <> map goMergingTree (toList t))
         PendingUnionMergeData ts ->
-          modelMerge MergeUnion (map goMergingTree ts)
+          modelMerge MR.MergeUnion (map goMergingTree ts)
 
     goPreExistingRun = \case
         PreExistingRunData r -> unRunData r
         PreExistingMergingRunData mr -> goMergingRun mr
 
-    goMergingRun :: IsMergeType t => SerialisedMergingRunData t -> Map SerialisedKey SerialisedEntry
+    goMergingRun :: MR.IsMergeType t => SerialisedMergingRunData t -> Map SerialisedKey SerialisedEntry
     goMergingRun = \case
         CompletedMergeData _ r -> unRunData r
         OngoingMergeData mt rs -> modelMerge mt (map (unRunData . toRunData) rs)
 
-modelMerge :: (Ord k, IsMergeType t) => t -> [Map k SerialisedEntry] -> Map k SerialisedEntry
+modelMerge :: (Ord k, MR.IsMergeType t) => t -> [Map k SerialisedEntry] -> Map k SerialisedEntry
 modelMerge mt = handleDeletes . Map.unionsWith (combine resolveVal)
   where
-    handleDeletes = if isLastLevel mt then Map.filter (/= Entry.Delete) else id
-    combine = if isUnion mt then Entry.combineUnion else Entry.combine
+    handleDeletes = if MR.isLastLevel mt then Map.filter (/= Entry.Delete) else id
+    combine = if MR.isUnion mt then Entry.combineUnion else Entry.combine
 
 resolveVal :: SerialisedValue -> SerialisedValue -> SerialisedValue
 resolveVal (SerialisedValue x) (SerialisedValue y) = SerialisedValue (x <> y)
+
+{-------------------------------------------------------------------------------
+  Supplying Credits
+-------------------------------------------------------------------------------}
+
+prop_supplyCredits ::
+     forall h.
+     FS.HasFS IO h
+  -> FS.HasBlockIO IO h
+  -> MR.CreditThreshold
+  -> NonEmpty MR.MergeCredits
+  -> MergingTreeData SerialisedKey SerialisedValue SerialisedBlob
+  -> IO Property
+prop_supplyCredits hfs hbio threshold credits mtd = do
+    FS.createDirectory hfs setupPath
+    FS.createDirectory hfs (FS.mkFsPath ["active"])
+    counter <- newUniqCounter 0
+    withMergingTree hfs hbio resolveVal runParams setupPath counter mtd $ \tree -> do
+      props <- for credits $ \c -> do
+        (MR.MergeDebt debt, _) <- remainingMergeDebt tree
+        leftovers <-
+          supplyCredits hfs hbio resolveVal runParams threshold root counter tree c
+        (MR.MergeDebt debt', _) <- remainingMergeDebt tree
+        return $
+          counterexample (show (debt, leftovers, debt')) $ conjoin [
+              counterexample "negative values" $
+                debt >= 0 && leftovers >= 0 && debt' >= 0
+            , counterexample "did not reduce debt sufficiently" $
+                debt' <= debt - (c - leftovers)
+            ]
+      return (conjoin (toList props))
```

Comment:
It might be nice to label test cases with the final debt, so that we can see whether tree merges ever complete